### PR TITLE
Make interface name configurable

### PR DESCRIPTION
**Pull request by plwhite**
Fixes issues #61 and #175.

As a result of changing some firewall code, I had to work on some previously untested code - and so this checkin enormously improves test coverage of invalid requests and responses in Felix, as well as tidying up that code.

### CODE REVIEW

**Comment by matthewdupre** [comment]
ok to test

**Comment by matthewdupre** [comment]
Two minor comments on the config file inline and one wider comment about specification of interface name format below.  The code itself looks fine.

Did you consider making the plugin entirely responsible for specifying interface naming?

Pros:
- Interface naming is orchestrator specific, and orchestrator specific code belongs in the plugin.
- No need for additional config options (since the correct settings will just be a lookup based on the plugin anyway).
- API is simplified by making the interface field compulsory.

Cons:
- We also use the prefix on its own for creating the intermediate chains.  This could also be passed on the API, but that would be a fairly unpleasant field; or refactored out of the rules, but that has implications of its own that I haven't fully thought through.
- Adds code to the plugins, but it's very simple.  They'd (other than whichever was lucky enough to be the default) all need equivalent instructions anyway.

We should talk about this and decide whether to make such a change.

**Comment by plwhite** [comment]
More changes.
- Interface prefix is specified on RESYNCSTATE response now.
- We make interface_name mandatory on the API.
- Get the suffix by taking the interface name and removing the prefix from it.
Back for review.

File: calico/felix/config.py [hunk 1]
```diff
@@ -217,4 +218,3 @@ def validate_addr(self, name, addr):
             raise ConfigException("Invalid or unresolvable %s value : %s" %
                                   (name, addr),
                                   self._config_path)
```
Comment (matthewdupre) [inline calico/felix/config.py]: Please remove this file from the change - we don't want the double blank line or to remove the blank line at end of file.

File: calico/felix/felix.py [hunk 2]
```diff
@@ -612,7 +686,14 @@ def run(self):
         endpoint_resync_needed = False
         acl_resync_needed = False
 
-        active_sockets = fsocket.poll(self.sockets.values(),
+        if self.iface_prefix:
+            poll_list = self.sockets.values()
+        else:
+            # Note got an first resync response (as no interface prefix), so
```
Comment (matthewdupre) [inline calico/felix/felix.py]: Note -> not.

File: calico/felix/test/data/felix_veth.cfg [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[global]
+FelixHostname = test_hostname
+# Plugin and ACL manager addresses
+PluginAddress = 127.0.0.1
+ACLAddress    = 127.0.0.1
+InterfacePrefix = veth
+
```
Comment (matthewdupre) [inline calico/felix/test/data/felix_veth.cfg]: InterfacePrefix isn't needed any more.

**Comment by matthewdupre** [comment]
I've made a few small comments above - also please update the API document and fix the UT failures above.

**Comment by plwhite** [comment]
Retest this

**Comment by Lukasa** [comment]
@neiljerram This looks like your tests are failing: care to take a look?

**Comment by neiljerram** [comment]
The UT code needs to be updated to expect the new interface_name key/value pair.

**Comment by Lukasa** [comment]
Sadly this pull request now requires rebasing. =(

**Pull request merged**